[ENH] `xLSTMTime` implementation

phoeenniixx opened the pull request:
### Description

This PR  tries to implement xLSTMTime based on this [paper](https://arxiv.org/pdf/2407.10240)

see also `sktime` issue [#6793](https://github.com/sktime/sktime/issues/6793)

### Checklist

- [x] Linked issues (if existing)
- [ ] Amended changelog for large changes (and added myself there as contributor)
- [ ] Added/modified tests
- [x] Used pre-commit hooks when committing to ensure that code is compliant with hooks. Install hooks with `pre-commit install`.
      To run hooks independent of commit, execute `pre-commit run --all-files`

phoeenniixx commented:
@fkiraly, I am new to `pytorch-forecasting` and its tests and all, can you please tell me exactly what am I "missing"?

phoeenniixx commented:
hi @fkiraly, I am new to `pytorch-forecasting` and its tests and all, can you please tell me exactly what am I "missing"?

phoeenniixx commented:
Will these tests suffice @fkiraly?

benHeid reviewed (changes requested):
Hi @phoeenniixx,
welcome to pytorch-forecasting and thank you for your pull request and contributing xlstm. 
I added first comments about the BaseClass you used. Please change it to one of the BaseClasses (see the comment). Since I suppose that this will change your code a bit. I will wait with a complete review until you changed it.

phoeenniixx commented:
Hi @benHeid, I need some help:

- here I implemented `xLSTMTime` class using `BaseModel` as for now I think this is the best fitted class... what do you think?

- Also, I made some changes in the `forward` function of the code where before it was accepting `Tensor` object, I changed it to `Dict` as I found out that the user mainly uses `TimeSeriesDataSet` and it returns a dict, please correct me if I am wrong here. 
-  I am using the `encoder_cont` key of the dict as input `x`. 

Please tell me if I am in a right direction
```python
class xLSTMTime(BaseModel):

    def __init__(
        self,
        input_size: int,
        hidden_size: int,
        output_size: int,
        xlstm_type: Literal['slstm', 'mlstm'],
        num_layers: int = 1,
        decomposition_kernel: int = 25,
        input_projection_size: Optional[int] = None,
        dropout: float = 0.1,
        loss: Metric = SMAPE(),
        device: Optional[torch.device] = None,
        **kwargs
    ):
        super().__init__(loss=loss, **kwargs)

        if xlstm_type not in ['slstm', 'mlstm']:
            raise ValueError("xlstm_type must be either 'slstm' or 'mlstm'")

        self.xlstm_type = xlstm_type
        self._device = device or torch.device('cuda' if torch.cuda.is_available() else 'cpu')
        self.to(self._device)

        self.decomposition = SeriesDecomposition(decomposition_kernel)
        self.batch_norm = nn.BatchNorm1d(hidden_size)

        self.input_projection_size = input_projection_size or hidden_size

        self.input_linear = None  

        if xlstm_type == 'mlstm':
            self.lstm = mLSTMNetwork(
                input_size=hidden_size,
                hidden_size=hidden_size,
                num_layers=num_layers,
                output_size=hidden_size,
                dropout=dropout,
                device=self.device
            )
        else:  # slstm
            self.lstm = sLSTMNetwork(
                input_size=hidden_size,
                hidden_size=hidden_size,
                num_layers=num_layers,
                output_size=hidden_size,
                dropout=dropout,
                device=self.device
            )

        self.output_linear = nn.Linear(hidden_size, output_size)
        self.instance_norm = nn.InstanceNorm1d(output_size)

    def forward(
        self,
        x: Dict[str, torch.Tensor],  
        hidden_states: Optional[
            Union[Tuple[torch.Tensor, torch.Tensor], Tuple[torch.Tensor, torch.Tensor, torch.Tensor]]
        ] = None
    ) -> Tuple[torch.Tensor, Union[Tuple[torch.Tensor, torch.Tensor], Tuple[torch.Tensor, torch.Tensor, torch.Tensor]]]:
   
        encoder_cont = x["encoder_cont"]
        batch_size, seq_len, n_features = encoder_cont.shape

        trend, seasonal = self.decomposition(encoder_cont)

        x = torch.cat([trend, seasonal], dim=-1)
        concatenated_features = x.shape[-1]

        if self.input_linear is None:
            self.input_linear = nn.Linear(concatenated_features, self.input_projection_size).to(self._device)

        x = self.input_linear(x)

        x = x.transpose(1, 2)  
        x = self.batch_norm(x)
        x = x.transpose(1, 2)  

        if hidden_states is None:
            hidden_states = self.lstm.init_hidden(batch_size)

        x = x.transpose(0, 1)
        output, hidden_states = self.lstm(x, *hidden_states)

        if isinstance(output, tuple):
            output = output[0]

        if output.dim() == 2:
            output = output.unsqueeze(0)
        output = self.output_linear(output)

        output = output.transpose(1, 2)
        output = self.instance_norm(output)
        output = output.transpose(1, 2)

        return output, hidden_states


    def predict(
            self,
            x: torch.Tensor,
            hidden_states: Optional[
                Union[Tuple[torch.Tensor, torch.Tensor], Tuple[torch.Tensor, torch.Tensor, torch.Tensor]]
            ] = None
    ) -> torch.Tensor:

        output, _ = self.forward(x, hidden_states)
        return output

    def training_step(self, batch, batch_idx):
        x, y = batch
        y = y[0] if isinstance(y, tuple) else y 

        y_pred, _ = self(x)

        if y_pred.ndim == 3 and y_pred.size(0) == 1:
            y_pred = y_pred.squeeze(0)  
        loss = self.loss(y_pred, y)
        self.log("train_loss", loss)
        return loss

    def validation_step(self, batch, batch_idx):
        x, y = batch
        y = y[0] if isinstance(y, tuple) else y 

        y_pred, _ = self(x)

        if y_pred.ndim == 3 and y_pred.size(0) == 1:
            y_pred = y_pred.squeeze(0)  
        loss = self.loss(y_pred, y)
        self.log("val_loss", loss)

        return loss




    def test_step(self, batch, batch_idx):
        x, y = batch
        y = y[0] if isinstance(y, tuple) else y 

        y_pred, _ = self(x)

        if y_pred.ndim == 3 and y_pred.size(0) == 1:
            y_pred = y_pred.squeeze(0)  
        loss = self.loss(y_pred, y)
        self.log("test_loss", loss)
        return loss

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)
        scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer, mode="min", factor=0.5, patience=10)
        return {"optimizer": optimizer, "lr_scheduler": scheduler, "monitor": "val_loss"}
```

phoeenniixx commented:
Also, Do we need to change the baseclass of just `xLSTMTime` only or `mLSTMNetwork` and `sLSTMNetwork` should also be changed?
(Although I think they are just a part of this main class so they could inherit from `nn.Module` without any problem?)

benHeid commented:
> * here I implemented `xLSTMTime` class using `BaseModel` as for now I think this is the best fitted class... what do you think?

Mhm. if the implementation does not support any exogenous features than either `BaseModel` or `AutoRegressiveBaseModel`. I would assume that the ladder is probably the better fit.

> * Also, I made some changes in the `forward` function of the code where before it was accepting `Tensor` object, I changed it to `Dict` as I found out that the user mainly uses `TimeSeriesDataSet` and it returns a dict, please correct me if I am wrong here.

I agree that a dict should be used here.

> * I am using the `encoder_cont` key of the dict as input `x`.

Yes that is the target time series.
 
> Please tell me if I am in a right direction

You might check the RNN implementation. Since this is also inheriting from an Autoregressive model and probably the most similar of the implemented models.
I would suggest that you check carefully, if you really need to implement the step / training_step method etc. or if is sufficient to use the inherited methods from the base class. 

But I think you are in the right direction.

phoeenniixx commented:
Hi @benHeid, I tried implementing `xLSTMTime` using `AutoRegressiveBaseModel` and I'm facing problem that some unexpected kwargs are being introduced here
implementation:
```python
from typing import Optional, Tuple, Union, Literal, Dict
import torch
from torch import nn
from pytorch_forecasting.metrics import SMAPE, Metric
from pytorch_forecasting.models.base_model import AutoRegressiveBaseModel
class xLSTMTime(AutoRegressiveBaseModel):

    def __init__(
        self,
        input_size: int,
        hidden_size: int,
        output_size: int,
        xlstm_type: Literal['slstm', 'mlstm'],
        num_layers: int = 1,
        decomposition_kernel: int = 25,
        input_projection_size: Optional[int] = None,
        dropout: float = 0.1,
        loss: Metric = SMAPE(),
        device: Optional[torch.device] = None,
        **kwargs
    ):
        self.save_hyperparameters()
        super().__init__(loss=loss, **kwargs)

        if xlstm_type not in ['slstm', 'mlstm']:
            raise ValueError("xlstm_type must be either 'slstm' or 'mlstm'")

        self.xlstm_type = xlstm_type
        self._device = device or torch.device('cuda' if torch.cuda.is_available() else 'cpu')
        self.to(self._device)

        self.decomposition = SeriesDecomposition(decomposition_kernel)
        self.batch_norm = nn.BatchNorm1d(hidden_size)

        self.input_projection_size = input_projection_size or hidden_size
        self.input_linear = None

        if xlstm_type == 'mlstm':
            self.lstm = mLSTMNetwork(
                input_size=hidden_size,
                hidden_size=hidden_size,
                num_layers=num_layers,
                output_size=hidden_size,
                dropout=dropout,
                device=self.device
            )
        else:  # slstm
            self.lstm = sLSTMNetwork(
                input_size=hidden_size,
                hidden_size=hidden_size,
                num_layers=num_layers,
                output_size=hidden_size,
                dropout=dropout,
                device=self.device
            )

        self.output_linear = nn.Linear(hidden_size, output_size)
        self.instance_norm = nn.InstanceNorm1d(output_size)

    def forward(
        self,
        x: Dict[str, torch.Tensor],
        hidden_states: Optional[
            Union[Tuple[torch.Tensor, torch.Tensor], Tuple[torch.Tensor, torch.Tensor, torch.Tensor]]
        ] = None,
    ) -> Dict[str, torch.Tensor]:



        encoder_cont = x["encoder_cont"]
        batch_size, seq_len, n_features = encoder_cont.shape

        trend, seasonal = self.decomposition(encoder_cont)


        x = torch.cat([trend, seasonal], dim=-1)
        concatenated_features = x.shape[-1]

        if self.input_linear is None:
            self.input_linear = nn.Linear(concatenated_features, self.input_projection_size).to(self.device)

        x = self.input_linear(x)

        x = x.transpose(1, 2)
        x = self.batch_norm(x)
        x = x.transpose(1, 2)

        if hidden_states is None:
            hidden_states = self.lstm.init_hidden(batch_size)

        x = x.transpose(0, 1)
        output, hidden_states = self.lstm(x, *hidden_states)

        if isinstance(output, tuple):
            output = output[0]

        if output.dim() == 2:
            output = output.unsqueeze(0)

        output = self.output_linear(output)

        output = output.transpose(1, 2)
        output = self.instance_norm(output)
        output = output.transpose(1, 2)

        output = output[..., :1]
        print(f"Output shape: {output.shape}")

        return self.to_network_output(prediction=output)

    def predict(
        self,
        x: torch.Tensor,
        hidden_states: Optional[
            Union[Tuple[torch.Tensor, torch.Tensor], Tuple[torch.Tensor, torch.Tensor, torch.Tensor]]
        ] = None
    ) -> torch.Tensor:

        output, _ = self.forward(x, hidden_states)
        return output
```


Code I'm using to test the working:

```python

import numpy as np
import pandas as pd
import torch
from pytorch_forecasting import TimeSeriesDataSet, TemporalFusionTransformer, Baseline
from lightning.pytorch import Trainer
from pytorch_forecasting.data import GroupNormalizer
from pytorch_lightning import seed_everything
from sklearn.model_selection import train_test_split

data = np.load("/content/pems-dataset/data/PEMS08/data.npy")  

n_time, n_nodes, n_features = data.shape
flattened_data = {
    "time_idx": np.tile(np.arange(n_time), n_nodes),
    "node_id": np.repeat(np.arange(n_nodes), n_time),
}
for feature_idx in range(n_features):
    flattened_data[f"feature_{feature_idx}"] = data[:, :, feature_idx].flatten()

flattened_df = pd.DataFrame(flattened_data)
flattened_df["target"] = flattened_df[f"feature_{n_features - 1}"]
flattened_df["node_id"] = flattened_df["node_id"].astype(str)


train_df, test_df = train_test_split(flattened_df, test_size=0.2, random_state=42)
train_df, val_df = train_test_split(train_df, test_size=0.2, random_state=42)


max_encoder_length = 30
max_prediction_length = 7
training = TimeSeriesDataSet(
    train_df,
    time_idx="time_idx",
    target="target",  
    group_ids=["node_id"],
    max_encoder_length=max_encoder_length,
    max_prediction_length=max_prediction_length,
    static_categoricals=["node_id"],
    time_varying_known_reals=[f"feature_{i}" for i in range(n_features)],
    time_varying_unknown_reals=["target"],
    target_normalizer=GroupNormalizer(groups=["node_id"]),
    allow_missing_timesteps=True,
)

validation = TimeSeriesDataSet.from_dataset(training, val_df)

batch_size = 64
train_dataloader = training.to_dataloader(train=True, batch_size=batch_size, num_workers=4)
val_dataloader = validation.to_dataloader(train=False, batch_size=batch_size, num_workers=4)


model_kwargs = {
    "input_size": len(training.reals),
    "hidden_size": 64,
    "output_size": 1,  
    "xlstm_type": "slstm",
    "num_layers": 2,
    "decomposition_kernel": 25,
    "dropout": 0.1,
    "loss": SMAPE(),
    "learning_rate": 1e-3,
}



model = xLSTMTime.from_dataset(training, **model_kwargs)

trainer = Trainer(
    max_epochs=30,
    gpus=1 if torch.cuda.is_available() else 0,
    gradient_clip_val=0.1,
)

trainer.fit(model, train_dataloaders=train_dataloader, val_dataloaders=val_dataloader)


test_data = TimeSeriesDataSet.from_dataset(training, test_df)
test_dataloader = test_data.to_dataloader(train=False, batch_size=batch_size, num_workers=4)

actuals = torch.cat([y[0] for x, y in iter(test_dataloader)])
predictions = model.predict(test_dataloader)

smape = SMAPE()(predictions, actuals)
print(f"Test SMAPE: {smape:.4f}")

```

Error:
```
TypeError: BaseModel.__init__() got an unexpected keyword argument 'target'
```

phoeenniixx commented:
I tried to solve this by deleting the specific kwargs that were unexpected before calling `super().__init()` in the class
```python

    if "target" in kwargs:
            del kwargs["target"]
    if "target_lags" in kwargs:
            del kwargs["target_lags"]
```
but if i remove these, then Trainer throws a similar unexpected keyword error
I think there is some problem with the test code, but I am not able to figure out what...
Can you please provide me with some insight.
Thank you!!

benHeid commented:
Sorry for my late response. Please ensure that the linting tests are green. Probably running the pre commit hooks locally should make it. 

Regarding the failing tests, you might check how the output currently looks like by manually executing the xLSTM. You might then see what the issue is. 

@fkiraly do we have any guides for pytorch-forecasting on how to write tests?

phoeenniixx commented:
Thanks for the reply @benHeid, actually the reason the tests are failing is: earlier I was using tensors, tuple etc and now TimeSeriesDataset is being used that uses a dict, that is the reason the tests are failing, I can correct those but I didn't do that because I noticed that for other models, they just use functions like `test_integration` etc. To write those functions, I first need to understand the input like `dataloaders`, `dataset` that is entered in these functions, like which data we are using here, the labels etc. is that data any arbitrary data or some pre-defined dataset?
Like look into this function from `test_models.test_rnn_model,py`:

```python
def _integration(
    data_with_covariates, tmp_path, cell_type="LSTM", data_loader_kwargs={}, clip_target: bool = False, **kwargs
):
    data_with_covariates = data_with_covariates.copy()
    if clip_target:
        data_with_covariates["target"] = data_with_covariates["volume"].clip(1e-3, 1.0)
    else:
        data_with_covariates["target"] = data_with_covariates["volume"]
    data_loader_default_kwargs = dict(
        target="target",
        time_varying_known_reals=["price_actual"],
        time_varying_unknown_reals=["target"],
        static_categoricals=["agency"],
        add_relative_time_idx=True,
    )
    data_loader_default_kwargs.update(data_loader_kwargs)
    dataloaders_with_covariates = make_dataloaders(data_with_covariates, **data_loader_default_kwargs)
    train_dataloader = dataloaders_with_covariates["train"]
    val_dataloader = dataloaders_with_covariates["val"]
    test_dataloader = dataloaders_with_covariates["test"]

    early_stop_callback = EarlyStopping(monitor="val_loss", min_delta=1e-4, patience=1, verbose=False, mode="min")

    logger = TensorBoardLogger(tmp_path)
    trainer = pl.Trainer(
        max_epochs=3,
        gradient_clip_val=0.1,
        callbacks=[early_stop_callback],
        enable_checkpointing=True,
        default_root_dir=tmp_path,
        limit_train_batches=2,
        limit_val_batches=2,
        limit_test_batches=2,
        logger=logger,
    )

    net = RecurrentNetwork.from_dataset(
        train_dataloader.dataset,
        cell_type=cell_type,
        learning_rate=0.15,
        log_gradient_flow=True,
        log_interval=1000,
        hidden_size=5,
        **kwargs,
    )
    net.size()
    try:
        trainer.fit(
            net,
            train_dataloaders=train_dataloader,
            val_dataloaders=val_dataloader,
        )
        test_outputs = trainer.test(net, dataloaders=test_dataloader)
        assert len(test_outputs) > 0
        # check loading
        net = RecurrentNetwork.load_from_checkpoint(trainer.checkpoint_callback.best_model_path)

        # check prediction
        net.predict(val_dataloader, fast_dev_run=True, return_index=True, return_decoder_lengths=True)
    finally:
        shutil.rmtree(tmp_path, ignore_errors=True)

    net.predict(val_dataloader, fast_dev_run=True, return_index=True, return_decoder_lengths=True)

```

Here they are using keys like "volume", and this is for `data_with_covariates` but I am not using the covariate base class that i can use directly this code and modify it to my requirements. I want to understand how this whole thing works and then I can write the test...

phoeenniixx commented:
for now I am just removing the test file and updating the code as required

fkiraly reviewed (changes requested):
Minor things before a more thorough review:

* can you kindly add tests for some basic use cases?
* can you make sure nothing except imports are in the `__init__` files? Similar to the recent change sin the repo.

phoeenniixx commented:
Thanks for the review @fkiraly!
A question:
to create the tests, I think `dataloaders_fixed_window_without_covariates` (of `tests.test_models.conftest.py`) should work?

phoeenniixx commented:
one more thing, I haven't added the docs for now as I wasn't sure which `basemodel` would be the best fit and will add those once it is clear.
For now I will add the docs for the `m_lstm` and `s_lstm`

fkiraly commented on pytorch_forecasting/layers/__init__.py:
```diff
@@ -12,18 +13,26 @@
 
 __all__ = [
     "FullAttention",
-    "TriangularCausalMask",
```
why does this line get removed?

phoeenniixx replied:
I didn't see any imports for `TriangularCausalMask`

phoeenniixx replied:
I actually found it in `layers._attention._full_attention`, it is not imported even in the `__init__` of `layers._attention`. I will add it to both the locations. At first, I thought it didnt exist 😅

phoeenniixx commented:
> the model should also be added to the model overview
I have seen some of the models having this docstring in `__init__` rather, so should we move those docstrings as well to the class?

fkiraly commented:
This PR is related: apparently not all models were properly tested or initializable via `__init__`: https://github.com/sktime/pytorch-forecasting/pull/1837

Do you know what the implicit contract is for `__init__`?

fkiraly commented:
minor comment regarding module structure - I would make it similar to other modules in naming:

* call the module `xlstm`
* the internal python module should be `_xlstm.py` etc

fkiraly commented on docs/source/models.rst:
```diff
@@ -31,6 +31,7 @@ and you should take into account. Here is an overview over the pros and cons of
    :py:class:`~pytorch_forecasting.models.deepar.DeepAR`,                                                  "x",          "x",                "x",          "",               "x",              "x",          "x [#deepvar]_ ",              "x",                       "",           3
    :py:class:`~pytorch_forecasting.models.temporal_fusion_transformer.TemporalFusionTransformer`,          "x",          "x",                "x",          "x",              "",               "x",          "",                            "x",                       "x",          4
    :py:class:`~pytorch_forecasting.model.tide.TiDEModel`,                                                  "x",          "x",                "x",          "",               "",               "",           "",                            "x",                       "",           3
+   :py:class:`~pytorch_forecasting.models.x_lstm_time.xLSTMTime`,                                          "x",          "x",                "x",          "",               "",               "",           "",                            "x",                       "",           3
```
this is incorrect now

phoeenniixx replied:
Oh sorry I forgot to change here

fkiraly reviewed (changes requested):
only minimal change requests.

* doclink is now broken due to rename
* could you move the lstm layers into a `_recurrent` folder, i.e., `layers._recurrent._mlstm` etc?

fkiraly reviewed (approved):
Great!
